### PR TITLE
add critical field to news flash

### PR DESCRIPTION
**Pull request by Shai2022**

### CODE REVIEW

File: anyway/models.py [hunk 1]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
+        suburban_road_severe_value=10,
+        suburban_road_killed_value=3,
+        urban_severe_value=2,
+    ):
+        from anyway.widgets.suburban_widgets.injured_count_by_severity_widget import (
+            InjuredCountBySeverityWidget,
+        )
+
+        if self.road1 is None or self.road_segment_name is None or self.date is None:
+            return None
```
Comment (atalyaalon) [inline anyway/models.py:899]: What happens in urban road? I think you might miss these news flashes

File: anyway/models.py [hunk 2]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
+        resolution = BE_CONST.ResolutionCategories(self.resolution)
+        five_years_ago = self.date.replace(year=self.date.year - years_before)
+        critical_values = InjuredCountBySeverityWidget.get_injured_count_by_severity(
+            self.road1, self.road_segment_name, five_years_ago, self.date
+        )
+        critical = False
+        if (
+            resolution == BE_CONST.ResolutionCategories.SUBURBAN_ROAD
+            or resolution == BE_CONST.ResolutionCategories.URBAN_JUNCTION
```
Comment (atalyaalon) [inline anyway/models.py:908]: Relevant urban resolution is STREET

File: anyway/models.py [hunk 3]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
+        ):
+            critical = (
+                (critical_values["severe_injured_count"] / suburban_road_severe_value)
+                + (critical_values["killed_count"] / suburban_road_killed_value)
+            ) >= 1
+        elif resolution == BE_CONST.ResolutionCategories.URBAN_JUNCTION:
```
Comment (atalyaalon) [inline anyway/models.py:914]: same here

**Review by atalyaalon — CHANGES_REQUESTED** [review]

File: anyway/models.py [hunk 4]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
+        urban_severe_value=2,
+    ):
+        from anyway.widgets.suburban_widgets.injured_count_by_severity_widget import (
+            InjuredCountBySeverityWidget,
+        )
+
+        if self.road1 is None or self.road_segment_name is None or self.date is None:
+            return None
+        resolution = BE_CONST.ResolutionCategories(self.resolution)
+        five_years_ago = self.date.replace(year=self.date.year - years_before)
```
Comment (atalyaalon) [inline anyway/models.py:901]: when using infographics data we use past 5 years slice last accident in production, and start from the first day of the year (January 1st)
I suggest doing the same and not using the newsflash date - 5 years

**Comment by atalyaalon** [comment]
@adirbru looks good! I added a few comments

File: anyway/models.py (hunk 2 above)
Comment (Shai2022) [reply anyway/models.py:908]: just to make sure i understand , there is two options:
SUBURBAN_ROAD
STREET
?

File: anyway/models.py (hunk 1 above)
Comment (Shai2022) [reply anyway/models.py:899]: how can i recognize urban road?

File: anyway/models.py (hunk 4 above)
Comment (Shai2022) [reply anyway/models.py:901]: ok, i will change this.

File: anyway/models.py (hunk 2 above)
Comment (atalyaalon) [reply anyway/models.py:908]: you are correct

File: anyway/models.py (hunk 1 above)
Comment (atalyaalon) [reply anyway/models.py:899]: both yishuv_name and street1_hebrew are not null

File: anyway/models.py (hunk 4 above)
Comment (atalyaalon) [reply anyway/models.py:901]: Thanks @Shai2022!

File: anyway/models.py [hunk 5]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
     newsflash_location_qualification = Column(
         Integer(),
         nullable=False,
         server_default=text(f"{NewsflashLocationQualification.NOT_VERIFIED.value}"),
     )
     location_qualifying_user = Column(BigInteger(), nullable=True)
 
+    def __init__(self, **kwargs):
+        super(NewsFlash, self).__init__(**kwargs)
+        self.critical = self.get_critical()
```
Comment (ziv17) [inline anyway/models.py:885]: Why do you choose to initialize this field in a different way from all other fields?

File: anyway/models.py [hunk 6]
```diff
@@ -871,13 +872,52 @@ class NewsFlash(Base):
...
+        suburban_road_killed_value=3,
+        urban_severe_value=2,
+    ):
+        from anyway.widgets.suburban_widgets.injured_count_by_severity_widget import (
+            InjuredCountBySeverityWidget,
+        )
+
+        if self.road1 is None or self.road_segment_name is None or self.date is None:
+            return None
+        resolution = BE_CONST.ResolutionCategories(self.resolution)
```
Comment (ziv17) [inline anyway/models.py:900]: I think it is worthwhile to handle the situation where the value in news_flash is not a legal `resolution`.

File: anyway/models.py (hunk 3 above)
Comment (ziv17) [inline anyway/models.py:914]: I think this `elif` will never be chosen as the `elif` above includes this possibility.

Comment (ziv17) [inline anyway/models.py:914]: I do not understand the testing for `URBAN_JUNCTION`. If the fields `road1` and `road_segment_name` are not null, then it is not urban. Ir it is `URBAN_JUNCTION`, then the location in this news_flash is erroneous.

**Review by ziv17 — COMMENTED** [review]
Nice work!

See a few comments below.

File: anyway/models.py (hunk 5 above)
Comment (Shai2022) [reply anyway/models.py:885]: 1. this field contains calculation and not the default value that given.
2. only when create a newflash recored
3. the super should init all other filed
4. only when write  a new flash

File: anyway/models.py (hunk 3 above)
Comment (Shai2022) [reply anyway/models.py:914]: i change the logic.

File: anyway/models.py (hunk 2 above)
Comment (Shai2022) [reply anyway/models.py]: done

File: anyway/models.py (hunk 6 above)
Comment (Shai2022) [reply anyway/models.py]: i init the critical field as None so if  type of the resolution, not SUBURBAN_ROAD or STREET 
i return None.
it this cover this case

File: anyway/models.py (hunk 3 above)
Comment (Shai2022) [reply anyway/models.py]: done

Comment (Shai2022) [reply anyway/models.py]: i change the logic, because what atalya said above.

File: anyway/models.py (hunk 5 above)
Comment (ziv17) [reply anyway/models.py:885]: I do not understand why we want to calculate the `critical` field each time we read from the DB. If it needs to be calculated periodically, then we can have a periodic process that updates it. The infographics code can read it from the DB like the other fields.

File: anyway/models.py (hunk 6 above)
Comment (ziv17) [reply anyway/models.py]: I am not sure. Maybe it is worthwhile to check in the news-flash table whether in these criteria the resolution is always legal.

File: anyway/models.py [hunk 7]
```diff
@@ -871,13 +872,54 @@ class NewsFlash(Base):
...
+            critical = (
+                (critical_values["severe_injured_count"] / suburban_road_severe_value)
+                + (critical_values["killed_count"] / suburban_road_killed_value)
+            ) >= 1
+        elif resolution == BE_CONST.ResolutionCategories.STREET:
+            critical = (
+                (critical_values["severe_injured_count"] / urban_severe_value)
+                + critical_values["killed_count"]
+            ) >= 1
+        return critical
```
Comment (ziv17) [inline anyway/models.py:921]: When `road1` and `road_segment` are not None, resolution will not be STREET. Anyway, even if it is, the `critical_values` here are calculated from `road1` and `road_segment`, and thus are not relevant to STREET locations.

File: anyway/models.py (hunk 5 above)
Comment (Shai2022) [reply anyway/models.py:885]: i only want to calc this value when a new newflash is created, there is a way to do this?
i might only calc this when a new newsflash is added, but i am not sure about the location of this in the code

File: anyway/models.py (hunk 7 above)
Comment (Shai2022) [reply anyway/models.py:921]: you right, i will talk with atalya on this issue and change the code after

File: anyway/models.py (hunk 6 above)
Comment (ziv17) [reply anyway/models.py]: > i init the critical field as None so if type of the resolution, not SUBURBAN_ROAD or STREET i return None. it this cover this case?

Note that the code that is used to calculate the values works only for STREET resolutions. If we need to calculate also for SUBURBAN_ROAD, we need to take the calculation from a different widget.

File: anyway/models.py (hunk 5 above)
Comment (ziv17) [reply anyway/models.py:885]: I think it is in function `scrape_all` in `anyway/parsers/news_flash.py`

**Comment by atalyaalon** [comment]
@Shai2022 can you make sure the critical field is added in new-flash API?
`localhost:8080/api/news-flash?limit=100&offset=0&resolution=suburban_road&resolution=street`

File: anyway/models.py (hunk 5 above)
Comment (Shai2022) [reply anyway/models.py]: i move the method to scrape_extract_store_rss and scrape_extract_store_twitter

**Comment by Shai2022** [comment]
@atalyaalon @ziv17 
after my talk with atalya i change the location of the set_critical value.
i addition i currenty added supoort for SUBURBAN_ROAD only.
for STREET support i think the widget InjuredAccidentsWithPedestriansWidget is the match one (@ziv17 correct me if i worng)
but i need to change the logic of the method to export this functionality, @atalyaalon i can to this in this pr or in a different one

**Comment by ziv17** [comment]
Hi @Shai2022 @atalyaalon ,
If our goal is to mark all new news-flash items that are critical, than maybe it makes sense to implement it with another approach, so it will be easier to maintain the code when the threshold for criticality will change:
My suggestion is to generate the infographics for this needed news-flash, and then go over the output, and test the results. We can have config or data file that will hold the widgets to test, and threshold for criticality. We can do it for each news-flash just before it is entered to the DB (by running the infographics on the news-flash location, or do a periodic process that go over all news-flash items that were not tested yet and calculates their criticality.

**Comment by atalyaalon** [comment]
@Shai2022 great work!

**Comment by atalyaalon** [comment]
@ziv17 I merged for now, since I want to add this to prod.
However, we can definitely modify the implementation. Since infographics already exists for any road segment / street, we can query the cache tables for the critical calculation. Is that your intention? This can be made at any stage (also in the stage of adding the news flash to the DB), since cache tables already there. What do you think?

@Shai2022 can you open a ticket for implementation of street resolution?

**Comment by ziv17** [comment]
No need to write the code that retrieves from cache, nor have different implementation per resolution.
It is possible to invoke one of the top level infographic functions with the location fields from te news-flash.

**Pull request merged**